### ISSUE TITLE
[DRAFT] SANS Q Resolution stored as generic function

### ISSUE THREAD

**Issue by wpotrzebowski** [scipp/esssans#111]
### Executive summary

Implement Q SANS resolution as a per q function and store it to NXCanSAS format

### Context and background knowledge

In time-of-flight SANS the resolution curves from all the wavelengths contributing to a particular Q value are averaged together, weighted by the number of neutrons detected at each wavelength. Since the resolution width is inversely proportional to wavelength this often gives a curve that is noticeably sharper in the peak and broader in the tails than a simple Gaussian.

Therefore, we need a q resolution description beyond a single value 

### Inputs

We need individual terms contributing to https://docs.mantidproject.org/nightly/algorithms/TOFSANSResolution-v1.html

### Methodology

The idea is to convolute resolution kernels to produce a function with arbitrary spacing

### Outputs

Save to NXCanSAS format. Placeholder /entry/data/q/resolutions and /entry/data/q/resolutions_descriptions

### Which interfaces are required?

Python module / function

### Test cases

At minimum, generated resolution function FWHM should be compared with dQ value from Milldener-Carpenter equation

**Comment by SimonHeybrock** [comment]
Most likely some differences, but it may be worth comparing to the implementation in https://github.com/scipp/essreflectometry/blob/main/src/essreflectometry/amor/resolution.py for reflectometry, at least to align naming, etc.

**Comment by nvaytet** [comment]
So in your equations above, $\sigma_{Q} / Q$ is dimensionless.

Now I looked again and it seems I made a mistake, the $1/\lambda^2$ is not multiplying everything, it is only the first part of the equation. So if you divide both sides of the equation by $Q^2$, then you are left with $1/(\lambda^2 Q^2)$ which is indeed dimensionless.

Regarding

> Also, the equation seems to differ from what is in the Mantid docs? There, I don't understand why they add the tof resolution to the wavelength resolution, it doesn't seem to make much sense. You either have your data as a function of wavelength or tof, not both at the same time?

I found https://docs.mantidproject.org/v3.9.0/algorithms/TOFSANSResolutionByPixel-v1.html which is more in agreement with the original equation you posted above?

**Comment by wpotrzebowski** [comment]
Yes, sorry I provided initially wrong link. It should this one https://docs.mantidproject.org/nightly/algorithms/TOFSANSResolutionByPixel-v1, which I belive provides the same equation as from v3.9.0

**Comment by olihammond** [comment]
> I found https://docs.mantidproject.org/v3.9.0/algorithms/TOFSANSResolutionByPixel-v1.html which is more in agreement with the original equation you posted above?

The implementation Neil mentioned above should be an appropriate baseline for our HC. Then, we simply need the output stored in the nxcansas as an additional dQ/Q column, for the corresponding bin centres.

**Comment by SimonHeybrock** [comment]
Extra info from chat: We may (at least initially) simply use the maximum (across pixels contributing to any given Q-bin) Q-resolution.

Note: Don't forget to exclude masked pixels.
